augmentation/augmentation.hpp

```diff
@@ -0,0 +1,182 @@
+/**
+ * @file augmentation.hpp
+ * @author Kartik Dutt
+ * 
+ * Definition of Augmentation class for augmenting data.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/methods/ann/layer/bilinear_interpolation.hpp>
+#include <boost/regex.hpp>
+
+#ifndef MODELS_AUGMENTATION_HPP
+#define MODELS_AUGMENTATION_HPP
+
+/**
+ * Augmentation class used to perform augmentations by transforming the data.
+ * For the list of supported augmentation, take a look at our wiki page.
+ *
+ * @code
+ * Augmentation augmentation({"horizontal-flip", "resize = (224, 224)"}, 0.2);
+ * augmentation.Transform(dataloader.TrainFeatures);
+ * @endcode
+ */
+class Augmentation
+{
+ public:
+  //! Create the augmentation class object.
+  Augmentation() :
+      augmentations(std::vector<std::string>()),
+      augmentationProbability(0.2)
+  {
+    // Nothing to do here.
+  }
+
+  /**
+   * Constructor for augmentation class.
+   *
+   * @param augmentations List of strings containing one of the supported
+   *                      augmentations.
+   * @param augmentationProbability Probability of applying augmentation on
+   *                                the dataset.
+   *                                NOTE : This doesn't apply to augmentations
+   *                                such as resize.
+   */
+  Augmentation(const std::vector<std::string>& augmentations,
+               const double augmentationProbability) :
+               augmentations(augmentations),
+               augmentationProbability(augmentationProbability)
+  {
+    // Sort the vector to place resize parameter to the front of the string.
+    // This prevents constant lookups for resize.
+    sort(this->augmentations.begin(), this->augmentations.end(), [](
```

Comment:
Would the augmentation names be case sensitive? Do you think they could be case-insensitive?

Reply:
Sure I can use the tolower util from mlpack. I'll add a commit fir it in the morning.